File: src/BOJ15650.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ15650 {
    // https://www.acmicpc.net/problem/15650

    // 변수 설정
    static FastReader fr = new FastReader();
    static int N, M;
    static int[] selected;
    static StringBuilder sb = new StringBuilder();

    // 입력 함수
    static void input(){
        N = fr.nextInt();
        M = fr.nextInt();
        selected = new int[M+1];
    }

    // 재귀함수 설계
    static void rec(int k){
        if(k > M){
            for(int index = 1; index < selected.length; index++){
                sb.append(selected[index] + " ");
            }
            sb.append('\n');
        } else {
            int start = k == 1 ? 1 : selected[k-1] + 1;
            for(int num = start; num <= N; num++){
                selected[k] = num;
                rec(k+1);
            }
        }
    }

    public static void main(String[] args) throws Exception {
        input();
        rec(1);

        System.out.println(sb);
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}

File: py/BOJ15650.py
import sys

input = sys.stdin.readline
# https://www.acmicpc.net/problem/15650

N, M = [int(s) for s in input().split()]
answer = [0] * (M + 1)


def permu(minimum, count):
    """
    재귀함수를 통해 순열의 각 자리수를 정한다.

    count : int
        몇번째 순열 요소인가.
    minimum : int
        현재 순열 순서에 사용될 수 있는 최소값
    """
    # 남은 숫자가 모자랄 경우
    if N - minimum < M - count:
        return
    # 다 선택하면 출력
    if count > M:
        print(" ".join([str(num) for num in answer[1:]]))
        return

    # 가능한 범위 내에서 선택하고 다음 자리로 재귀 호출
    for num in range(minimum, N + 1):
        if num > answer[count - 1]:
            answer[count] = num
            permu(num + 1, count + 1)


permu(1, 1)
